src/coreclr/jit/target.h

```diff
@@ -785,45 +778,112 @@ inline SingleTypeRegSet genRegMaskFloat(regNumber reg ARM_ARG(var_types type /*
-// genRegMask: Given a register, and its type, generate the appropriate regMask
+// genSingleTypeRegMask: Given a register, generate the appropriate regMask
+//
+// Arguments:
+//    regNum   - the register of interest
+//
+// Return Value:
+//    This will usually return the same value as genRegMask(regNum), except
+//    that it will return a 64-bits (or 32-bits) entity instead of `regMaskTP`.
+//
+inline SingleTypeRegSet genSingleTypeRegMask(regNumber reg)
+{
+    assert((unsigned)reg < ArrLen(regMasks));
+#ifdef TARGET_AMD64
+    // shift is faster than a L1 hit on modern x86
+    // (L1 latency on sandy bridge is 4 cycles for [base] and 5 for [base + index*c] )
+    // the reason this is AMD-only is because the x86 BE will try to get reg masks for REG_STK
+    // and the result needs to be zero.
+    SingleTypeRegSet result = 1ULL << reg;
+    assert(result == regMasks[reg]);
+    return result;
+#else
+    return regMasks[reg];
+#endif
+}
+
+//------------------------------------------------------------------------
+// genSingleTypeRegMask: Given a register, generate the appropriate regMask
 //
 // Arguments:
 //    regNum   - the register of interest
 //    type     - the type of regNum (i.e. the type it is being used as)
 //
 // Return Value:
-//    This will usually return the same value as genRegMask(regNum), but
-//    on architectures where multiple registers are used for certain types
+//    This will usually return the same value as genRegMask(regNum), except
+//    that it will return a 64-bits (or 32-bits) entity instead of `regMaskTP`.
+//    On architectures where multiple registers are used for certain types
 //    (e.g. TYP_DOUBLE on ARM), it will return a regMask that includes
-//    all the registers.
-//    Registers that are used in pairs, but separately named (e.g. TYP_LONG
-//    on ARM) will return just the regMask for the given register.
-//
-// Assumptions:
-//    For registers that are used in pairs, the caller will be handling
-//    each member of the pair separately.
+//    all the registers for that type.
 //
-inline SingleTypeRegSet genRegMask(regNumber regNum, var_types type)
+inline SingleTypeRegSet genSingleTypeRegMask(regNumber regNum, var_types type)
 {
 #if defined(TARGET_ARM)
     SingleTypeRegSet regMask = RBM_NONE;
 
     if (varTypeUsesIntReg(type))
     {
-        regMask = genRegMask(regNum);
+        regMask = genSingleTypeRegMask(regNum);
     }
     else
     {
         assert(varTypeUsesFloatReg(type));
-        regMask = genRegMaskFloat(regNum, type);
+        regMask = genSingleTypeFloatMask(regNum, type);
     }
 
     return regMask;
 #else
-    return genRegMask(regNum);
+    return genSingleTypeRegMask(regNum);
 #endif
 }
 
+/*****************************************************************************
+ *
+ *  Map a register number to a register mask.
+ */
+
+inline regMaskTP genRegMask(regNumber reg)
+{
+    // TODO: Populate regMaskTP based on reg
+    return genSingleTypeRegMask(reg);
```

Comment:
Even with predicate registers I would expect us to end up with basically the same logic as before, i.e. something like `regMaskTP(1) << reg`, probably encapsulating that operation inside `regMaskTP`. That is, we leave the `regNumber` mappings for predicate registers to just follow the float registers so that all reg numbers are unique.

Of course LSRA is going to need slightly different logic to map `SingleTypeRegSet` to `regMaskTP` which ends up requiring additional context about the type of set.